### PR TITLE
CREATE debugging continued

### PR DESCRIPTION
**Pull request by OlivierBBB**

### CODE REVIEW

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/AccountSnapshot.java [hunk 1]
```diff
@@ -245,6 +245,15 @@ public AccountSnapshot raiseNonceByOne() {
...
 
+  public AccountSnapshot setDeploymentNumber(Hub hub) {
+    return this.setDeploymentNumber(hub.transients.conflation().deploymentInfo());
+  }
+
+  public AccountSnapshot setDeploymentNumber(DeploymentInfo deploymentInfo) {
+    this.deploymentNumber(deploymentInfo.deploymentNumber(address));
+    return this;
+  }
+
```
Comment (OlivierBBB) [inline arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/AccountSnapshot.java:256]: This is important and reflects a subtle mistake. When we "undo" a row we want to
- copy over the `DEPLOYMENT_STATUS` (criss-cross style)
- get the current `DEPLOYMENT_NUMBER`

The `setDeploymentInfo` method overwrites the deployment status. As such it's unsuitable for **undoing** operations.

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/fragment/StackFragment.java [hunk 2]
```diff
@@ -109,7 +109,9 @@ private StackFragment(
           hub.currentFrame().frame().getStackItem(0).toUnsignedBigInteger();
       final BigInteger codeSize = BigInteger.valueOf(hub.currentFrame().code().getSize());
 
-      boolean prospectivePcNewIsInBounds = codeSize.compareTo(prospectivePcNew) > 0;
+      boolean noOutOfGasException = !Exceptions.outOfGasException(exceptions);
+      boolean prospectivePcNewIsInBounds =
+          codeSize.compareTo(prospectivePcNew) > 0 && noOutOfGasException;
```
Comment (OlivierBBB) [inline arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/fragment/StackFragment.java:122]: Genuine mistake: `OOGX` has precedence in exception testing for `JUMP(I)` instructions over `JUMPX`

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/fragment/common/CommonFragment.java [hunk 3]
```diff
@@ -117,8 +117,8 @@ public Trace trace(Trace trace) {
         .gasCost(Bytes.ofUnsignedLong(commonFragmentValues.gasCostToTrace()))
         .gasNext(
             Bytes.ofUnsignedLong(isExec && isUnexceptional() ? commonFragmentValues.gasNext : 0))
-        .refundCounter(commonFragmentValues.gasRefund)
-        .refundCounterNew(commonFragmentValues.gasRefundNew)
+        .refundCounter((commonFragmentValues.hubProcessingPhase == TX_EXEC) ? commonFragmentValues.gasRefund : 0)
+        .refundCounterNew((commonFragmentValues.hubProcessingPhase == TX_EXEC) ? commonFragmentValues.gasRefundNew : 0)
```
Comment (OlivierBBB) [inline arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/fragment/common/CommonFragment.java]: `REFUND` columns only print during execution phase (`TX_EXEC`)

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/fragment/imc/mmu/MmuCall.java [hunk 4]
```diff
@@ -261,7 +261,7 @@ public static MmuCall invalidCodePrefix(final Hub hub) {
                     .frame()
                     .shadowReadMemory(0, hub.currentFrame().frame().memoryByteSize())))
         .sourceOffset(EWord.of(hub.messageFrame().getStackItem(0)))
-        .successBit(!Exceptions.invalidCodePrefix(currentExceptions));
+        .successBit(Exceptions.invalidCodePrefix(currentExceptions));
```
Comment (OlivierBBB) [inline arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/fragment/imc/mmu/MmuCall.java:264]: removing double negation

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/call/CallSection.java [hunk 5]
```diff
@@ -163,8 +163,10 @@ public CallSection(Hub hub, MessageFrame frame) {
 
     if (Exceptions.any(exceptions)) {
       scenarioFragment.setScenario(CALL_EXCEPTION);
-      final XCallOobCall oobCall = new XCallOobCall();
-      firstImcFragment.callOob(oobCall);
+      if (opCode == CALL) {
+        final XCallOobCall oobCall = new XCallOobCall();
+        firstImcFragment.callOob(oobCall);
+      }
```
Comment (OlivierBBB) [inline arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/call/CallSection.java:169]: Only `CALL` among all CALL-type instructions requires the `OOB_INST_XCALL` OOB instruction.

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/halt/ReturnSection.java [hunk 6]
```diff
@@ -140,7 +146,7 @@ public ReturnSection(Hub hub, MessageFrame messageFrame) {
       final MmuCall actuallyInvalidCodePrefixMmuCall = MmuCall.invalidCodePrefix(hub);
       firstImcFragment.callMmu(actuallyInvalidCodePrefixMmuCall);
 
-      checkArgument(!actuallyInvalidCodePrefixMmuCall.successBit());
+      checkArgument(actuallyInvalidCodePrefixMmuCall.successBit());
```
Comment (OlivierBBB) [inline arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/halt/ReturnSection.java:147]: Removing double negation stuff again

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/halt/ReturnSection.java [hunk 7]
```diff
@@ -230,7 +238,7 @@ public ReturnSection(Hub hub, MessageFrame messageFrame) {
       firstImcFragment.callOob(maxCodeSizeOobCall);
 
       // sanity checks
-      checkArgument(invalidCodePrefixCheckMmuCall.successBit());
+      checkArgument(!invalidCodePrefixCheckMmuCall.successBit());
```
Comment (OlivierBBB) [inline arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/halt/ReturnSection.java:238]: Again, double negation stuff

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/halt/ReturnSection.java [hunk 8]
```diff
@@ -299,10 +301,10 @@ public void resolveUponRollback(Hub hub, MessageFrame messageFrame, CallFrame ca
         hub.factories()
             .accountFragment()
             .make(
-                postDeploymentAccountSnapshot,
-                undoingDeploymentAccountSnapshot,
+                    firstCreateeNew,
+                    secondCreateeNew,
                 DomSubStampsSubFragment.revertWithCurrentDomSubStamps(
-                    this.hubStamp(), hub.callStack().currentCallFrame().revertStamp(), 1));
+                    hubStamp, hub.callStack().currentCallFrame().revertStamp(), 1));
```
Comment (OlivierBBB) [inline arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/halt/ReturnSection.java:304]: We must use the `hubStamp` of the time of creation of the `ReturnSection`.

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/halt/ReturnSection.java [hunk 9]
```diff
@@ -61,31 +63,35 @@ public class ReturnSection extends TraceSection
...
-  AccountSnapshot undoingDeploymentAccountSnapshot;
+  AccountSnapshot firstCreatee;
+  AccountSnapshot firstCreateeNew;
+  AccountSnapshot secondCreatee;
+  AccountSnapshot secondCreateeNew;
   ContextFragment squashParentContextReturnData;
   Address deploymentAddress;
 
+  int hubStamp;
+
```
Comment (OlivierBBB) [inline arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/halt/ReturnSection.java:72]: Required to correctly set the `SUB_STAMP` for reverted deployments.

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/halt/ReturnSection.java [hunk 10]
```diff
@@ -61,31 +63,35 @@ public class ReturnSection extends TraceSection
   final ReturnScenarioFragment returnScenarioFragment;
   AccountFragment deploymentFragment;
 
-  AccountSnapshot preDeploymentAccountSnapshot;
-  AccountSnapshot postDeploymentAccountSnapshot;
-  AccountSnapshot undoingDeploymentAccountSnapshot;
+  AccountSnapshot firstCreatee;
+  AccountSnapshot firstCreateeNew;
+  AccountSnapshot secondCreatee;
+  AccountSnapshot secondCreateeNew;
```
Comment (OlivierBBB) [inline arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/halt/ReturnSection.java:67]: Just sticking to the spec.

File: arithmetization/src/test/java/net/consensys/linea/UnitTestWatcher.java [hunk 11]
```diff
@@ -22,6 +22,6 @@
 public class UnitTestWatcher extends LineaTestWatcher {
   @Override
   public String getTestName(ExtensionContext context) {
-    return context.getDisplayName();
+    return context.getRequiredTestMethod().getName() + ",   " + context.getDisplayName();
```
Comment (OlivierBBB) [inline arithmetization/src/test/java/net/consensys/linea/UnitTestWatcher.java:25]: Better display names for parametric tests.

File: arithmetization/src/test/java/net/consensys/linea/zktracer/instructionprocessing/EmptyDeploymentsInTheRootTest.java [hunk 12]
```diff
@@ -63,24 +63,37 @@ public class EmptyDeploymentsInTheRootTest {
...
           .push(0) // size
-          .push(0) // offset
+          .push(0x0f) // offset
           .op(OpCode.REVERT)
           .compile();
 
   final Bytes initCodeNonemptyRevert =
       BytecodeCompiler.newProgram()
-          .op(OpCode.BLOCKHASH) // value
+          .op(OpCode.COINBASE) // value
```
Comment (OlivierBBB) [inline arithmetization/src/test/java/net/consensys/linea/zktracer/instructionprocessing/EmptyDeploymentsInTheRootTest.java:72]: `BLOCKHASH` on its own produces a `stackUnderflowException` 😅. This was not the goal. I have extracted the `BLOCKHASH` case into a separate test.

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/CreateSection.java [hunk 13]
```diff
@@ -64,17 +65,23 @@ public class CreateSection extends TraceSection
...
+  private AccountSnapshot secondCreateeNew;
 
-  // Just at the entry child frame
-  private AccountSnapshot reEntryCreatorSnapshot;
-  private AccountSnapshot reEntryCreateeSnapshot;
+  // Second account rows
+  private AccountSnapshot thirdCreator;
+  private AccountSnapshot thirdCreatee;
+  private AccountSnapshot thirdCreatorNew;
+  private AccountSnapshot thirdCreateeNew;
```
Comment (OlivierBBB) [inline arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/CreateSection.java:84]: Just a big refactoring of `CreateSection`. The goal being to be close to the spec.

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/CreateSection.java [hunk 14]
```diff
@@ -87,11 +94,14 @@ public class CreateSection extends TraceSection
   private Wei value;
   private boolean success = false;
 
+  private int hubStamp;
+
```
Comment (OlivierBBB) [inline arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/CreateSection.java:98]: Required for RevertStamp business.

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/halt/ReturnSection.java [hunk 15]
```diff
@@ -290,7 +291,8 @@ public void resolveUponRollback(Hub hub, MessageFrame messageFrame, CallFrame ca
             ? RETURN_FROM_DEPLOYMENT_NONEMPTY_CODE_WILL_REVERT
             : RETURN_FROM_DEPLOYMENT_EMPTY_CODE_WILL_REVERT);
 
-    undoingDeploymentAccountSnapshot = AccountSnapshot.canonical(hub, deploymentAddress);
+    secondCreatee = firstCreateeNew.deepCopy().setDeploymentNumber(hub);
+    secondCreateeNew = firstCreatee.deepCopy().setDeploymentNumber(hub);
```
Comment (OlivierBBB) [inline arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/halt/ReturnSection.java:292]: No more "taking snapshots at revert time". This was the root cause for shit going sideways all the time.

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/romlex/RomLex.java [hunk 16]
```diff
@@ -189,7 +187,11 @@ public void callRomLex(final MessageFrame frame) {
         checkArgument(length > 0, "callRomLex for RETURN expects positive size");
 
         byteCode = frame.shadowReadMemory(offset, length);
-        hub.defers().scheduleForContextExit(this, hub.currentFrame().id());
+        Address deploymentAddress = hub.currentFrame().byteCodeAddress();
+        final ContractMetadata contractMetadata =
+                ContractMetadata.make(deploymentAddress, hub.deploymentNumberOf(deploymentAddress), false);
+        final RomOperation chunk = new RomOperation(contractMetadata, false, true, byteCode);
+        operations.add(chunk);
```
Comment (OlivierBBB) [inline arithmetization/src/main/java/net/consensys/linea/zktracer/module/romlex/RomLex.java:194]: Why were we scheduling production of the contract metadata etc ? We have all the information required.

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/CreateSection.java [hunk 17]
```diff
@@ -250,148 +258,149 @@ public void resolveUponContextEntry(Hub hub, MessageFrame frame) {
...
-    reEntryCreatorSnapshot = childContextEntryCreatorSnapshot.deepCopy().incrementBalanceBy(value);
-    reEntryCreateeSnapshot =
-        childContextEntryCreateeSnapshot
-            .deepCopy()
-            .decrementBalanceBy(value)
-            .deploymentStatus(false)
-            .deploymentNumber(hub.deploymentNumberOf(createeAddress))
-            .nonce(0);
+        secondCreator = firstCreatorNew.deepCopy().setDeploymentNumber(hub);
+        secondCreatorNew = firstCreatorNew.deepCopy().incrementBalanceBy(value);
```
Comment (OlivierBBB) [inline arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/CreateSection.java]: Clean up house

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/fragment/common/CommonFragment.java [hunk 18]
```diff
@@ -117,8 +117,14 @@ public Trace trace(Trace trace) {
...
-        .refundCounter(commonFragmentValues.gasRefund)
-        .refundCounterNew(commonFragmentValues.gasRefundNew)
+        .refundCounter(
+            (commonFragmentValues.hubProcessingPhase == TX_EXEC)
+                ? commonFragmentValues.gasRefund
+                : 0)
+        .refundCounterNew(
+            (commonFragmentValues.hubProcessingPhase == TX_EXEC)
+                ? commonFragmentValues.gasRefundNew
+                : 0)
```
Comment (OlivierBBB) [inline arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/fragment/common/CommonFragment.java:127]: Only print refunds during the `TX_EXEC` phase

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/fragment/imc/ImcFragment.java [hunk 19]
```diff
@@ -137,7 +137,14 @@ public Trace trace(Trace trace) {
...
     for (TraceSubFragment subFragment : moduleCalls) {
-      subFragment.trace(trace, hub.state.stamps());
+      if (subFragment instanceof MmuCall) {
+        MmuCall mmuCall = (MmuCall) subFragment;
+        if (mmuCall.traceMe()) {
+          subFragment.trace(trace, hub.state.stamps());
+        }
+      } else {
+        subFragment.trace(trace, hub.state.stamps());
+      }
```
Comment (OlivierBBB) [inline arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/fragment/imc/ImcFragment.java:147]: This fixes an issue where the HUB would trace the MMU instruction for reverted LOG's

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/AccountSection.java [hunk 20]
```diff
@@ -37,11 +36,15 @@ public class AccountSection extends TraceSection implements PostRollbackDefer {
...
-  AccountSnapshot accountSnapshotAfter;
+  AccountSnapshot firstAccountSnapshot;
+  AccountSnapshot firstAccountSnapshotNew;
+  AccountSnapshot secondAccountSnapshot;
+  AccountSnapshot secondAccountSnapshotNew;
+  int hubStamp;
 
   public AccountSection(Hub hub) {
     super(hub, maxNumberOfRows(hub));
+    hubStamp = hub.stamp();
```
Comment (OlivierBBB) [inline arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/AccountSection.java:47]: Same old rationalizing:
- the AccountSnapshots are closer to the spec
- we log the `hub.stamp()` for later reverts

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/AccountSection.java [hunk 21]
```diff
@@ -87,29 +90,30 @@ public AccountSection(Hub hub) {
...
-    final AccountSnapshot preRollBackAccountSnapshot =
-        accountSnapshotAfter.deepCopy().setDeploymentInfo(hub);
-    final AccountSnapshot postRollBackAccountSnapshot =
-        accountSnapshotBefore.deepCopy().setDeploymentInfo(hub);
+    secondAccountSnapshot = firstAccountSnapshotNew.deepCopy().setDeploymentNumber(hub);
+    secondAccountSnapshotNew = firstAccountSnapshot.deepCopy().setDeploymentNumber(hub);
     final DomSubStampsSubFragment undoingDomSubStamps =
         DomSubStampsSubFragment.revertWithCurrentDomSubStamps(
-            this.hubStamp(), hub.currentFrame().revertStamp(), 1);
+            hubStamp, hub.currentFrame().revertStamp(), 1);
```
Comment (OlivierBBB) [inline arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/AccountSection.java:111]: This fixes the `revertDomSubStamps`

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/halt/ReturnSection.java [hunk 22]
```diff
@@ -316,23 +315,27 @@ public void resolvePostTransaction(
...
             .make(
-                preDeploymentAccountSnapshot,
-                postDeploymentAccountSnapshot,
-                DomSubStampsSubFragment.standardDomSubStamps(this.hubStamp(), 0));
+                firstCreatee,
+                firstCreateeNew,
+                DomSubStampsSubFragment.standardDomSubStamps(hubStamp, 0));
+
+    if (nonemptyByteCode) {
+      deploymentAccountFragment.requiresRomlex(true);
```
Comment (OlivierBBB) [inline arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/halt/ReturnSection.java:337]: Omission

File: arithmetization/src/test/java/net/consensys/linea/zktracer/exceptions/MemoryExpansionExceptionTest.java [hunk 23]
```diff
@@ -54,6 +55,21 @@ public void memoryExpansionExceptionTest(boolean triggerRoob, OpCode opCode) {
...
+    new MxpTestUtils().triggerNonTrivialButMxpxOrRoobForOpCode(program, triggerRoob, opCode);
+    BytecodeRunner bytecodeRunner = BytecodeRunner.of(program.compile());
+    bytecodeRunner.run();
+    assertEquals(
+        MEMORY_EXPANSION_EXCEPTION,
+        bytecodeRunner.getHub().previousTraceSection().commonValues.tracedException());
+    assertTrue(bytecodeRunner.getHub().mxp().operations().getLast().getMxpCall().isMxpx());
+    assertEquals(triggerRoob, bytecodeRunner.getHub().mxp().operations().getLast().isRoob());
+  }
+
```
Comment (OlivierBBB) [inline arithmetization/src/test/java/net/consensys/linea/zktracer/exceptions/MemoryExpansionExceptionTest.java:72]: ras: just to test a particular test vector.

**Review by gusiri — APPROVED** [review]

**Pull request merged**